net/proxy_resolution/proxy_config_service_tor.cc

```diff
@@ -0,0 +1,283 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/net/proxy_resolution/proxy_config_service_tor.h"
+
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <sstream>
+#include <vector>
+
+#include "base/no_destructor.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/string_util.h"
+#include "base/time/time.h"
+#include "base/values.h"
+#include "crypto/random.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+#include "net/proxy_resolution/proxy_config_with_annotation.h"
+#include "net/proxy_resolution/proxy_resolution_service.h"
+#include "url/origin.h"
+
+namespace net {
+
+namespace {
+
+constexpr NetworkTrafficAnnotationTag kTorProxyTrafficAnnotation =
+    DefineNetworkTrafficAnnotation("proxy_config_tor", R"(
+      semantics {
+        sender: "Proxy Config Tor"
+        description:
+          "Establishing a connection through the tor proxy server"
+        trigger:
+          "Whenever a network request is made from a tor profile."
+        data:
+          "Proxy configuration."
+        destination: OTHER
+        destination_other:
+          "The proxy server specified in the configuration."
+      }
+      policy {
+        cookies_allowed: NO
+        setting: "This feature cannot be disabled by settings."
+        policy_exception_justification: "Not implemented."
+      })");
+
+static base::NoDestructor<std::map<
+    ProxyResolutionService*,
+    ProxyConfigServiceTor::TorProxyMap>> tor_proxy_map_;
+
+}  // namespace
+
+const int kTorPasswordLength = 16;
+// Default tor circuit life time is 10 minutes
+constexpr base::TimeDelta kTenMins = base::TimeDelta::FromMinutes(10);
+
+ProxyConfigServiceTor::TorProxyMap* GetTorProxyMap(
+    ProxyResolutionService* service) {
+  return &(tor_proxy_map_.get()->operator[](service));
```

Comment:
DCHECK if not found?

Reply:
this is effectively a lazy initializer because of the value type https://github.com/brave/brave-core/pull/2790/files#diff-99dc16b687b8e58fa3b90a0b455c6331R51 so it will never be not found. If you want to check for existence without creating new entry then you have to use find or an iterator https://github.com/brave/brave-core/pull/2790/files#diff-99dc16b687b8e58fa3b90a0b455c6331R120

Reply:
oh yes, right.